homeassistant/components/kef/media_player.py

```diff
@@ -0,0 +1,274 @@
+    SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP,
+    MediaPlayerDevice,
+)
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PORT,
+    CONF_TYPE,
+    STATE_OFF,
+    STATE_ON,
+)
+from homeassistant.helpers import config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "KEF"
+DEFAULT_PORT = 50001
+DEFAULT_MAX_VOLUME = 0.5
+DEFAULT_VOLUME_STEP = 0.05
+DEFAULT_INVERSE_SPEAKER_MODE = False
+
+DOMAIN = "kef"
+
+SCAN_INTERVAL = timedelta(seconds=30)
+PARALLEL_UPDATES = 0
+
+SOURCES = {"LSX": ["Wifi", "Bluetooth", "Aux", "Opt"]}
+SOURCES["LS50"] = SOURCES["LSX"] + ["Usb"]
+
+SUPPORT_KEF = (
+    SUPPORT_VOLUME_SET
+    | SUPPORT_VOLUME_STEP
+    | SUPPORT_VOLUME_MUTE
+    | SUPPORT_SELECT_SOURCE
+    | SUPPORT_TURN_OFF
+    | SUPPORT_TURN_ON
+)
+
+CONF_MAX_VOLUME = "maximum_volume"
+CONF_VOLUME_STEP = "volume_step"
+CONF_INVERSE_SPEAKER_MODE = "inverse_speaker_mode"
+CONF_STANDBY_TIME = "standby_time"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_TYPE): vol.In(["LS50", "LSX"]),
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_MAX_VOLUME, default=DEFAULT_MAX_VOLUME): cv.small_float,
+        vol.Optional(CONF_VOLUME_STEP, default=DEFAULT_VOLUME_STEP): cv.small_float,
+        vol.Optional(
+            CONF_INVERSE_SPEAKER_MODE, default=DEFAULT_INVERSE_SPEAKER_MODE
+        ): cv.boolean,
+        vol.Optional(CONF_STANDBY_TIME): vol.In([20, 60]),
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the KEF platform."""
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+
+    host = config[CONF_HOST]
+    speaker_type = config[CONF_TYPE]
+    port = config[CONF_PORT]
+    name = config[CONF_NAME]
+    maximum_volume = config[CONF_MAX_VOLUME]
+    volume_step = config[CONF_VOLUME_STEP]
+    inverse_speaker_mode = config[CONF_INVERSE_SPEAKER_MODE]
+    standby_time = config.get(CONF_STANDBY_TIME)
+
+    sources = SOURCES[speaker_type]
+
+    _LOGGER.debug(
+        "Setting up %s with host: %s, port: %s, name: %s, sources: %s",
+        DOMAIN,
+        host,
+        port,
+        name,
+        sources,
+    )
+
+    try:
+        if ipaddress.ip_address(host).version == 6:
+            mode = "ip6"
+        else:
+            mode = "ip"
+    except ValueError:
+        mode = "hostname"
+    mac = await hass.async_add_executor_job(partial(get_mac_address, **{mode: host}))
+
+    unique_id = f"kef-{mac}"
+
+    media_player = KefMediaPlayer(
+        name,
+        host,
+        port,
+        maximum_volume,
+        volume_step,
+        standby_time,
+        inverse_speaker_mode,
+        sources,
+        ioloop=hass.loop,
+        unique_id=unique_id,
+    )
+
+    if unique_id in hass.data[DOMAIN]:
```

Comment:
It's probably better to use host as instance lifetime id, since unique_id might be `None`.

Reply:
Fixed in 5db883d6fd653451d6235c008d8b653bad9b1aac.